### ISSUE TITLE
`(:A)-[:r]->(:B) ≠  (:A)-[:r*1..1]->(:B)`

### ISSUE THREAD

**Issue by phf-1** [kuzudb/kuzu#3262]
Hello,

Thank you for your great work.
While testing Kuzu, I found that: `(:A)-[:r]->(:B) ≠  (:A)-[:r*1..1]->(:B)`
Is it supposed to be the case?

Cheers,
PHF

----

Here is a way to reproduce this observation, modulo a few renamings:
```Python
import kuzu
import shutil

tmpdir = "/tmp/kuzu_test"
shutil.rmtree(tmpdir)
db = kuzu.Database(tmpdir)
conn = kuzu.Connection(db)
queries = [
    'CREATE NODE TABLE Exercise (xxx String, PRIMARY KEY (xxx));',
    'CREATE REL TABLE next (FROM Exercise TO Exercise, xxx String);',
    'CREATE NODE TABLE Source (xxx String, PRIMARY KEY (xxx));',
    'CREATE NODE TABLE User (xxx String, PRIMARY KEY (xxx));',
    'CREATE REL TABLE GROUP has (FROM User TO Source, FROM Source TO Exercise, xxx String);',
    'CREATE (ex:Exercise {xxx: "11-ex"});',
    'CREATE (ex:Exercise {xxx: "12-ex"});',
    'CREATE (ex:Exercise {xxx: "13-ex"});',
    'MATCH (ex1:Exercise {xxx: "11-ex"}), (ex2:Exercise {xxx: "12-ex"}) CREATE (ex1)-[:next {xxx: "next"}]->(ex2);',
    'MATCH (ex1:Exercise {xxx: "12-ex"}), (ex2:Exercise {xxx: "13-ex"}) CREATE (ex1)-[:next {xxx: "next"}]->(ex2);',
    'CREATE (src:Source {xxx: "1-src"});',
    'MATCH (src:Source {xxx: "1-src"}), (ex:Exercise {xxx: "11-ex"}) CREATE (src)-[:has {xxx: "has"}]->(ex);',
    'MATCH (src:Source {xxx: "1-src"}), (ex:Exercise {xxx: "12-ex"}) CREATE (src)-[:has {xxx: "has"}]->(ex);',
    'MATCH (src:Source {xxx: "1-src"}), (ex:Exercise {xxx: "13-ex"}) CREATE (src)-[:has {xxx: "has"}]->(ex);',
    'CREATE (usr:User {xxx: "usr"});',
    'MATCH (usr:User {xxx: "usr"}), (src:Source {xxx: "1-src"}) CREATE (usr)-[:has {xxx: "has"}]->(src);',
    'CREATE (ex:Exercise {xxx: "21-ex"});',
    'CREATE (ex:Exercise {xxx: "22-ex"});',
    'MATCH (ex1:Exercise {xxx: "21-ex"}), (ex2:Exercise {xxx: "22-ex"}) CREATE (ex1)-[:next {xxx: "next"}]->(ex2);',
    'CREATE (src:Source {xxx: "2-src"});',
    'MATCH (src:Source {xxx: "2-src"}), (ex:Exercise {xxx: "21-ex"}) CREATE (src)-[:has {xxx: "has"}]->(ex);',
    'MATCH (src:Source {xxx: "2-src"}), (ex:Exercise {xxx: "22-ex"}) CREATE (src)-[:has {xxx: "has"}]->(ex);',
    'MATCH (usr:User {xxx: "usr"}), (src:Source {xxx: "2-src"}) CREATE (usr)-[:has {xxx: "has"}]->(src);',
]

for query in queries:
    conn.execute(query, {})
```

![Screenshot from 2024-04-12 10-48-13](https://github.com/kuzudb/kuzu/assets/166803203/3a20cf7e-f7c7-4e2b-95e2-f57aecdfb349)
![Screenshot from 2024-04-12 10-48-39](https://github.com/kuzudb/kuzu/assets/166803203/1ab85611-ae76-4b15-9b23-10238259e9bf)

**Comment by prrao87** [comment]
Hi @phf-1, I'd say that this is intended behaviour.

Consider a simpler example with our [users dataset](https://github.com/kuzudb/kuzu/tree/master/dataset/demo-db/csv).
```py
import kuzu

db = kuzu.Database("./persons")
conn = kuzu.Connection(db)

res1 = conn.execute("MATCH (a:User)-[r:Follows]->(b:User) RETURN a.name, b.name, r.since")
while res1.has_next():
    print(res1.get_next())
```

When you specify `-[r:Follows]->`, the expectation is that you return only the relationships, not paths.
This is what you'd get:
```
['Adam', 'Karissa', 2020]
['Adam', 'Zhang', 2020]
['Karissa', 'Zhang', 2021]
['Zhang', 'Noura', 2022]
```

On the other hand, specifying the `*` operator as `-[r:Follows*1..1]->` indicates to Kùzu that you want to return *paths*, obtained via recursive joins.

```py
res2 = conn.execute("MATCH (a:User)-[r:Follows*1..1]->(b:User) RETURN a.name, b.name, r")
while res2.has_next():
    print(res2.get_next())
```

When the paths are output, it contains an offset value that points to the location of the relationship in the variable length path (you only find out the length of the path at runtime):
```
['Karissa', 'Zhang', {'_nodes': [], '_rels': [{'_src': {'offset': 1, 'table': 0}, '_dst': {'offset': 2, 'table': 0}, '_label': 'Follows', 'since': 2021}]}]
['Adam', 'Karissa', {'_nodes': [], '_rels': [{'_src': {'offset': 0, 'table': 0}, '_dst': {'offset': 1, 'table': 0}, '_label': 'Follows', 'since': 2020}]}]
['Adam', 'Zhang', {'_nodes': [], '_rels': [{'_src': {'offset': 0, 'table': 0}, '_dst': {'offset': 2, 'table': 0}, '_label': 'Follows', 'since': 2020}]}]
['Zhang', 'Noura', {'_nodes': [], '_rels': [{'_src': {'offset': 2, 'table': 0}, '_dst': {'offset': 3, 'table': 0}, '_label': 'Follows', 'since': 2022}]}]
```

The length of the result in both cases is the same (whether or not the recursive join operator is used). However, recursive joins are expected to return paths containing relationships and their offsets, not relationships alone. Does this make sense?

@andyfengHKU feel free to clarify anything if you think it doesn't.

**Comment by prrao87** [comment]
Actually, now I see a slight discrepancy @andyfengHKU 😅 .

I should be outputting the paths from the first query too:

```py
res1 = conn.execute("MATCH (a:User)-[r:Follows]->(b:User) RETURN a.name, b.name, r")
while res1.has_next():
    print(res1.get_next())
```

The offset values are different - why is this?
```
['Adam', 'Karissa', {'_src': {'offset': 0, 'table': 0}, '_dst': {'offset': 1, 'table': 0}, '_label': 'Follows', 'since': 2020}]
['Adam', 'Zhang', {'_src': {'offset': 0, 'table': 0}, '_dst': {'offset': 2, 'table': 0}, '_label': 'Follows', 'since': 2020}]
['Karissa', 'Zhang', {'_src': {'offset': 1, 'table': 0}, '_dst': {'offset': 2, 'table': 0}, '_label': 'Follows', 'since': 2021}]
['Zhang', 'Noura', {'_src': {'offset': 2, 'table': 0}, '_dst': {'offset': 3, 'table': 0}, '_label': 'Follows', 'since': 2022}]
```

**Comment by prrao87** [comment]
This seems related to this other issue:
https://github.com/kuzudb/kuzu/issues/3169

**Comment by andyfengHKU** [comment]
@phf-1 thanks for reporting. I'll take a look.

**Comment by andyfengHKU** [comment]
Hi @phf-1 this should be fixed in PR #3384
